n2n.7
.TH "n2n_v3" 7 "Sep 27, 2021" "version 3" "Background"
.SH NAME
n2n version 3 \- version 3 of the n2n decentralised peer-to-peer network overlay
VPN.
.SH DESCRIPTION
n2n is a peer-to-peer network overlay or VPN system that provides layer 2 over
layer 3 encapsulation with data transform capabilities such as encryption and
compression. This guide also discusses the differences of version 3 of n2n from
version 2.
.SH PROTOCOLS
n2n-3 basically uses the same set of messages to communicate with edges and
supernodes. However, due to slight packet format changes, the n2n-3 messages
are not compatible with n2n-2. There is no backward compatibility for n2n-2.
.SH ENCRYPTION
n2n-3 offers four different ciphers for payload encryption as well as optional
header encryption. Earlier versions of n2n-2 provided a mechanism using a key
schedule which has been removed in n2n-3. A basic user authentication scheme
relying on asymmetric cryptography has been added to n2n-3.

n2n-3 provides the following ciphers to chose from for payload encryption; more
can be added as required:
.TP
.B (1) NULL
Data is encapsulated unchanged. Useful for testing and high-performance, low
sensitivity applications.
.TP
.B (2) TF-CTS
Twofish AES candidate in CTS mode.
.TP
.B (3) AES-CTS
AES in CTS mode with up to 256-bit key.
.TP
.B (4) CHACHA20
ChaCha20, a well known stream cipher developped by Daniel J. Bernstein.
.TP
.B (5) SPECK-CTR
A fast block cipher developped by the NSA used as stream cipher in CTR mode.
.TP
Full Header Encyption
The optional full header encryption also encrypts packets' header which include
some administrative data. In addition, it adds replay protection.
.TP
User Password Authentication
n2n-3 implements an optional user-password authentication scheme. A key
generator assists in generating user's public keys to be stored at the
supernode side.
.SH COMPRESSION
LZO for payload compression is an always available option at n2n-3. If compiled with
zstdlib support, ZSTD is at optional service as well.
.SH EXTENSIBILITY
n2n-3 decouples the data transform system from the core of the edge
operation. This allows for easier addition of new data transform
operations. n2n-3 reserves some standard transform identifiers (such as TwoFish
encryption) but allocates transform identifiers for user-defined
transforms. This allows anyone to add to n2n new private transforms without
breaking compatibility with the standard offering.
.SH FEDERATED SUPERNODES
n2n-3 incorporates the capability of multiple supernodes to be federated.
Federation acts transparently and balances the workload evenly among the
federated supernodes. Supernodes keep track of edges connected to different
supernodes and forward packets as required. This feature naturally supports
fail-over and this increases redundancy and resilience.
.P
Information on additional supernodes is propagated to all edges. In addition,
the n2n-3 edge implementation allows multiple supernodes to be specified on the
command line. Edges monitor the current supernode for responses to
REGISTER_SUPER as well as PING messages. After three responses from current
supernode are missed or when a better supernode in terms of significant lower workload
is found, the edge tries to connect to another supernode. It cycles through the list
f supernodes which over and over again is sorted according to reported workload.

.SH MANAGEMENT CONSOLE
Edge and supernode in n2n-3 provide a UDP-based management console. Both listen
on the localhost address 127.0.0.1. Commands can be sent to the programs by
sending to the UDP socket. Responses are returned to the socket from which
commands were issued. This only works from the computer on which the programs
are running. Statistics can be retrieved and commands issued. The netcat utility
is all that is required; but more sophisticated tools could be built on the
interface.

.SH SUPERNODE AUTHENTICATION
The supernode federation name serves as private key shared between the supernodes only.
The corresponding public key can be provided to the edges.

.SH MESSAGE SUMMARY
The following message types work within n2n-3.
.TP
REGISTER_SUPER
Sent from an edge to its local supernode to register its MAC with the community.
Also, federated supernodes use this packet format to register to each other.
.TP
REGISTER_SUPER_ACK
Sent from a supernode to an edge to confirm registration. This also carries the
definition of the edge socket as seen at the supernode so NAT can be detected
and described. Furthermore, it carries information about additional federated
supernodes.
.TP
REGISTER_SUPER_NAK
Supernode refusing to register an edge.
.TP
PACKET
Encapsulated ethernet packets sent between edges. Supernodes forward or
broadcast these and edges send them direct in peer-to-peer mode.
.TP
REGISTER
A peer-to-peer mode registration request from one edge to another. Supernodes
forward these to facilitate NAT crossing introductions.
.TP
REGISTER_ACK
Complete peer-to-peer mode setup between two edges. These messages need to
travel direct between edges.
.TP
QUERY_PEER
Queries a supernode about another edge, especially its public socket in case of
no peer-to-peer communication can be established. Additionally, it serves as PING
to query supernodes about themselves.
.TP
PEER_INFO
Answers the QUERY_PEER; it also covers the special case of the PING query, internally
called PONG.
.SH AUTHORS
.TP
Richard Andrews andrews (at) ntop.org - main author of n2n-2
.TP
Luca Deri
deri (at) ntop.org - code inherited from n2n-1
.SH SEE ALSO
ifconfig(8) edge(8) supernode(1)
.br
the documentation contained in the source code
.br
the extensive documentation found in n2n's \fBdoc/\fR folder


CHANGELOG.md
# Changelog

## n2n 2.8 (August 2020)

This release brings significant new features to n2n's crypto world and offers
some compression opportunities. The added support for routing table manipulation
might increase comfort. Besides further honing existing features, this release
addresses some bugs.

### New Features

* Two lightweight stream ciphers: ChaCha20 (optional, through OpenSSL) & SPECK (integrated)
* Full Header Encryption (including packet checksumming as well as replay protection)
* A callback interface to better integrate n2n in third party software (you can still use it stand-alone)
* Enable the integrated LZO1x compression
* Add optional ZSTD compression (through zstdlib)
* Support for changing system routes at program start and end
* User and group id parameter for supernode
* Application of cryptography in n2n is seperately documented
* Add a new pseudo random number generator with higher periodicity seeded with more entropy if available

### Improvements

* Have AES and ChaCha20 use OpenSSL's `evp_*` interface to make better use of available hardware acceleration
* Fix invalid sendto when supernode name resolution fails
* Update to supernode's purge logic
* Extended management supernode's port output
* Fix read tap device failed when OS wakes up from sleep
* Free choice of supernode's management UDP port (for multiple supernodes on one machine)
* Additional trace messages to better indicate established connections and connection type
* Fix edge's register-to-supernode loop
* Remove redundant code
* Restructure the code in directories
* Clean-up platform-dependant code
* Compile fixes for Windows
* Fix build warnings
* …and many more under-the-hood fixes and tunings

## n2n 2.6 (March 2020)

The 2.6 release is mostly a maintenance release to address the issues 
of 2.4 that has been the first release since a long time of silence.

### New Features

* AES encryption that features an overall speed bump (12x speed) and security with respect to Twofish used in the previous n2n version
* Add ability to specify a whitelist of allowed communities on the supernode
* Implement local peers discovery via multicast
* Full peer-to-peer topology support.
* Add support for multiple edge systemd services
* Add benchmark tool for the encryption throughput
* Implement packet stats for P2P vs supernode communication
* Automatically drop privileges to user n2n
* Add support for ARM64 build
* More options to control MTU, P2P connections, TOS and log verbosity
* Implement a wireshark dissector for the n2n protocol
* Implement n2n-decode utility to decode and dump traffic to PCAP


### Improvements
* Extensive Windows and OpenWRT support.
* Windows compilation fixes and instructions
* Instructions and makefile file to build n2n on OpenWRT
* MacOS compilation fixes and instructions
* Improve the connection stability and the chances to establish a P2P connection
* Stable and more resilient connection.
* Remove keyschedule support to simplify the encryption code
* Replace peers linked list with hash table for faster lookup in big networks
* Integrate the changes made in the meyerd fork of n2n
* Remove calls to system() in tuntap_linux and use netlink instead
* n2n version improvements

## n2n 2.4 (August 2018)

This is the first release after 2012 and thus it is focusing mainly
on making it work on current operating system versions, so that the
next release will be based on modern code.

### New Features
* Added deb/rpm packages
* Added systemd configuration files
* Added ability to read configuration files instead of using only the CLI (needed for packaging)
* Added n2n Android app
* Implemented simple API to embed n2n in applications (in addition to use it stand-alone)

### Improvements
* Major code cleanup
* Fixed compilation issues on MacOS
* Fixed Linux segmentation fault


wireshark/n2n.lua
-- (C) 2019 - ntop.org and contributors

n2n = Proto("n2n", "n2n Protocol")

-- #############################################

PKT_TYPE_PING               = 0
PKT_TYPE_REGISTER           = 1
PKT_TYPE_DEREGISTER         = 2
PKT_TYPE_PACKET             = 3
PKT_TYPE_REGISTER_ACK       = 4
PKT_TYPE_REGISTER_SUPER     = 5
PKT_TYPE_REGISTER_SUPER_ACK = 6
PKT_TYPE_REGISTER_SUPER_NAK = 7
PKT_TYPE_FEDERATION         = 8
PKT_TYPE_PEER_INFO          = 9
PKT_TYPE_QUERY_PEER         = 10

PKT_TRANSFORM_NULL      = 1
PKT_TRANSFORM_TWOFISH   = 2
PKT_TRANSFORM_AESCBC    = 3

FLAG_FROM_SUPERNODE   = 0x0020
FLAG_SOCKET           = 0x0040
FLAG_OPTIONS          = 0x0080

SOCKET_FAMILY_AF_INET  = 0x0000
SOCKET_FAMILY_AF_INET6 = 0x8000

-- #############################################

version = ProtoField.uint8("n2n.version", "version", base.DEC)
ttl = ProtoField.uint8("n2n.ttl", "ttl", base.DEC)

packet_type_mask = 0x001f
pkt_type_2_str = {
  [PKT_TYPE_PING] = "ping",
  [PKT_TYPE_REGISTER] = "register",
  [PKT_TYPE_DEREGISTER] = "deregister",
  [PKT_TYPE_PACKET] = "packet",
  [PKT_TYPE_REGISTER_ACK] = "register_ack",
  [PKT_TYPE_REGISTER_SUPER] = "register_super",
  [PKT_TYPE_REGISTER_SUPER_ACK] = "register_super_ack",
  [PKT_TYPE_REGISTER_SUPER_NAK] = "register_super_nak",
  [PKT_TYPE_FEDERATION] = "federation",
  [PKT_TYPE_PEER_INFO] = "peer_info",
  [PKT_TYPE_QUERY_PEER] = "query_peer",
}
packet_type = ProtoField.uint8("n2n.packet_type", "packetType", base.HEX, pkt_type_2_str, packet_type_mask)

flags_mask = 0xffe0
flags = ProtoField.uint16("n2n.flags", "Flags", base.HEX, nil, flags_mask)
from_supernode_flag = ProtoField.uint16("n2n.flags.from_supernode", "from_supernode", base.BOOLEAN, nil, FLAG_FROM_SUPERNODE)
socket_flag = ProtoField.uint16("n2n.flags.socket", "socket", base.BOOLEAN, nil, FLAG_SOCKET)
options_flag = ProtoField.uint16("n2n.flags.options", "options", base.BOOLEAN, nil, FLAG_OPTIONS)
community = ProtoField.string("n2n.community", "Community", base.ASCII)

-- #############################################

src_mac = ProtoField.ether("n2n.src_mac", "Source")
dst_mac = ProtoField.ether("n2n.dst_mac", "Destination")
socket_info = ProtoField.none("n2n.socket", "Socket Info")
socket_family = ProtoField.uint16("n2n.socket.family", "Family", base.HEX, {
  [0] = "AF_INET",
})
socket_port = ProtoField.uint16("n2n.socket.port", "Port")
socket_ipv4 = ProtoField.ipv4("n2n.socket.ipv4", "IPv4")
socket_ipv6 = ProtoField.ipv6("n2n.socket.ipv6", "IPv6")

-- #############################################

peer_info_field = ProtoField.none("n2n.peer_info", "PeerInfo")
peer_info_flags = ProtoField.uint16("n2n.peer_info.flags", "Flags")
peer_info_mac = ProtoField.ether("n2n.peer_info.query_mac", "Query")

query_peer_field = ProtoField.none("n2n.query_peer", "QueryPeer")

-- #############################################



packet_field = ProtoField.none("n2n.packet", "Packet")
packet_transform = ProtoField.uint16("n2n.packet.transform", "Transform", base.HEX, {
  [PKT_TRANSFORM_NULL] = "Plaintext",
  [PKT_TRANSFORM_TWOFISH] = "TwoFish",
  [PKT_TRANSFORM_AESCBC] = "AES CBC",
})
packet_payload = ProtoField.bytes("n2n.packet.payload", "Payload")

-- #############################################

register_field = ProtoField.none("n2n.register", "Register")
register_cookie = ProtoField.uint32("n2n.register.cookie", "Cookie", base.HEX)

register_ack_field = ProtoField.none("n2n.register_ack", "RegisterACK")
register_ack_cookie = ProtoField.uint32("n2n.register_ack.cookie", "Cookie", base.HEX)

register_super_field = ProtoField.none("n2n.register_super", "RegisterSuper")
register_super_cookie = ProtoField.uint32("n2n.register_super.cookie", "Cookie", base.HEX)
register_super_auth_schema = ProtoField.uint16("n2n.register_super.auth.schema", "AuthSchema", base.HEX)
register_super_auth_data = ProtoField.uint16("n2n.register_super.auth.data", "AuthData", base.HEX)

register_super_ack_field = ProtoField.none("n2n.register_super_ack", "RegisterSuperACK")
register_super_ack_cookie = ProtoField.uint32("n2n.register_super_ack.cookie", "Cookie", base.HEX)
register_super_ack_lifetime = ProtoField.uint16("n2n.register_super_ack.lifetime", "Registration Lifetime", base.DEC)
register_super_ack_num_sn = ProtoField.uint8("n2n.register_super_ack.num_sn", "Num Supernodes", base.DEC)

-- #############################################

n2n.fields = {
  version, ttl, packet_type,
  flags, from_supernode_flag, socket_flag, options_flag,
  community,

  -- Generic
  src_mac, dst_mac,
  socket_info, socket_family, socket_port, socket_ipv4, socket_ipv6,

  -- PKT_TYPE_REGISTER
  register_field, register_cookie,
  -- PKT_TYPE_PACKET
  packet_field, packet_transform, packet_payload,
  -- PKT_TYPE_REGISTER_ACK
  register_ack_field, register_ack_cookie,
  -- PKT_TYPE_REGISTER_SUPER
  register_super_field, register_super_cookie, register_super_auth_schema, register_super_auth_data,
  -- PKT_TYPE_REGISTER_SUPER_ACK
  register_super_ack_field, register_super_ack_cookie, register_super_ack_lifetime, register_super_ack_num_sn,
  -- PKT_TYPE_PEER_INFO
  peer_info_field, peer_info_flags, peer_info_mac,
  -- PKT_TYPE_QUERY_PEER
  query_peer_field,
}

-- #############################################

function dissect_socket(subtree, buffer, offset)
  local sock_baselen = 4
  local sock_protolen = 0
  buffer = buffer(offset)
  local sock_family = bit.band(buffer(0,4):uint(), 0xFFFF0000)

  if(sock_family == SOCKET_FAMILY_AF_INET) then
    sock_protolen = 4
  elseif(sock_family == SOCKET_FAMILY_AF_INET6) then
    sock_protolen = 16
  end

  local totlen = sock_baselen + sock_protolen
  local socktree = subtree:add(socket_info, buffer(0, totlen))

  socktree:add(socket_family, buffer(0, 2))
  socktree:add(socket_port, buffer(2, 2))

  if(sock_family == SOCKET_FAMILY_AF_INET) then
    socktree:add(socket_ipv4, buffer(4, sock_protolen))
  elseif(sock_family == SOCKET_FAMILY_AF_INET6) then
    socktree:add(socket_ipv6, buffer(4, sock_protolen))
  end

  return offset+totlen, socktree
end

-- #############################################

function dissect_register(subtree, buffer, flags)
  local regtree = subtree:add(register_field, buffer)

  regtree:add(register_cookie, buffer(0,4))
  regtree:add(src_mac, buffer(4,6))
  regtree:add(dst_mac, buffer(10,6))

  if(bit.band(flags, FLAG_SOCKET) == FLAG_SOCKET) then
    dissect_socket(regtree, buffer, 16)
  end

  return regtree
end

-- #############################################

function dissect_register_ack(subtree, buffer, flags)
  local regtree = subtree:add(register_ack_field, buffer)

  regtree:add(register_ack_cookie, buffer(0,4))
  regtree:add(src_mac, buffer(4,6))
  regtree:add(dst_mac, buffer(10,6))

  return regtree
end

-- #############################################

function dissect_packet(subtree, buffer, flags, pinfo)
  local pktree = subtree:add(packet_field, buffer)

  pktree:add(src_mac, buffer(0,6))
  pktree:add(dst_mac, buffer(6,6))

  if(bit.band(flags, FLAG_SOCKET) == FLAG_SOCKET) then
    idx = dissect_socket(pktree, buffer, 12)
  else
    idx = 12
  end

  pktree:add(packet_transform, buffer(idx,2))
  local payload = pktree:add(packet_payload, buffer(idx+2))
  local transform = buffer(idx,2):uint()

  -- Can only dissect unencrypted data
  if(transform == PKT_TRANSFORM_NULL) then
    Dissector.get("eth_withoutfcs"):call(buffer(idx+2):tvb(), pinfo, payload)
  end

  return pktree
end

-- #############################################

function dissect_register_super(subtree, buffer, flags)
  local regtree = subtree:add(register_super_field, buffer)

  regtree:add(register_super_cookie, buffer(0,4))
  regtree:add(src_mac, buffer(4,6))
  regtree:add(register_super_auth_schema, buffer(10,2))
  regtree:add(register_super_auth_data, buffer(12,2))

  return regtree
end

-- #############################################

function dissect_register_super_ack(subtree, buffer, flags)
  local regtree = subtree:add(register_super_ack_field, buffer)

  regtree:add(register_super_ack_cookie, buffer(0,4))
  regtree:add(dst_mac, buffer(4,6))
  regtree:add(register_super_ack_lifetime, buffer(10,2))
  local idx = dissect_socket(regtree, buffer, 12)
  regtree:add(register_super_ack_num_sn, buffer(idx, 1))

  return regtree
end

-- #############################################

function dissect_peer_info(subtree, buffer, flags)
  local peertree = subtree:add(peer_info_field, buffer)

  peertree:add(peer_info_flags, buffer(0,2))
  peertree:add(peer_info_mac, buffer(2,6))
  dissect_socket(peertree, buffer, 8)

  return peertree
end

-- #############################################

function dissect_query_peer(subtree, buffer, flags)
  local peertree = subtree:add(query_peer_field, buffer)

  peertree:add(src_mac, buffer(0,6))
  peertree:add(dst_mac, buffer(6,6))

  return peertree
end

-- #############################################

function n2n.dissector(buffer, pinfo, tree)
  local length = buffer:len()
  if length < 20 then return end

  pinfo.cols.protocol = n2n.name

  local pkt_type = bit.band(buffer(2,2):uint(), packet_type_mask)
  local subtree = tree:add(n2n, buffer(), string.format("n2n Protocol, Type: %s", pkt_type_2_str[pkt_type] or "Unknown"))

  -- Common
  subtree:add(version, buffer(0,1))
  subtree:add(ttl, buffer(1,1))
  subtree:add(packet_type, buffer(2,2))
  local flags_buffer = buffer(2,2)
  local flags_tree = subtree:add(flags, flags_buffer)
  subtree:add(community, buffer(4,16))

  -- Flags
  flags_tree:add(from_supernode_flag, flags_buffer)
  flags_tree:add(socket_flag, flags_buffer)
  flags_tree:add(options_flag, flags_buffer)

  -- Packet specific
  local flags = bit.band(buffer(2,2):uint(), flags_mask)
  local typebuf = buffer(20)

  if(pkt_type == PKT_TYPE_REGISTER) then
    dissect_register(subtree, typebuf, flags)
  elseif(pkt_type == PKT_TYPE_REGISTER) then
    dissect_register_ack(subtree, typebuf, flags)
  elseif(pkt_type == PKT_TYPE_PACKET) then
    dissect_packet(subtree, typebuf, flags, pinfo)
  elseif(pkt_type == PKT_TYPE_REGISTER_SUPER) then
    dissect_register_super(subtree, typebuf, flags)
  elseif(pkt_type == PKT_TYPE_REGISTER_SUPER_ACK) then
    dissect_register_super_ack(subtree, typebuf, flags)
  elseif(pkt_type == PKT_TYPE_PEER_INFO) then
    dissect_peer_info(subtree, typebuf, flags)
  elseif(pkt_type == PKT_TYPE_QUERY_PEER) then
    dissect_query_peer(subtree, typebuf, flags)
  end
end

-- #############################################

local udp_port = DissectorTable.get("udp.port")
udp_port:add(50001, n2n)
